### ISSUE TITLE
Many Datastore Namespaces

### ISSUE THREAD

**Issue by rajesh99** [GoogleCloudPlatform/DataflowJavaSDK#83]
Following up on the following issue. Since the following issue is closed, so I am creating a new issue.
https://github.com/GoogleCloudPlatform/DataflowJavaSDK/issues/64

I have many namespaces in the cloud datastore.  The DatastoreWordCount example shows using the pipeline to read and write to one datastore namespace.
What should be done to read/write data to thousands of namespaces.  Should we generate thousands
of pipelines in a loop for each namespace?
What is the recommended approach?

**Comment by dhalperi** [comment]
I think that is roughly correct. You should not combine entities from different namespaces in the same pipeline. Making any call to the Datastore API that wants to query/write entities from different namespaces should fail.

Instead, suppose my pipeline for a single namespace is something like this:

```java
public void handleOneNamespace(Pipeline p, Query query, String dataset, String namespace) {
    DatastoreIO.source source = DatastoreIO.source()
        .withQuery(query)
        .withDataset(dataset)
        .withNamespace(namespace);
    p.apply(Read.from(source))
     .apply(ParDo.of("transform " + namespace, ...))
     .apply(DatastoreIO.writeTo(dataset));
    return;
}
```

and I have a `List<String>` containing all the namespaces I want to process, then I can run:

```java
Query q = ...;
String dataset = ...;
List<String> namespaces = ...;
Pipeline p = ...;
for (String namespace : namespaces) {
    handleOneNamespace(p, query, dataset, namespace);
}
p.run();
```

The resulting Dataflow job would process every namespace in parallel, without every processing entities from different namespaces together.

However, you will need to experiment with how well this scales. If you have 1000s of namespace I would expect it to fail. I might start with processing namespaces in blocks of 10.

**Comment by chandra-vs** [comment]
Even i have the same issue. I have more than 5000 namespaces, as you said it may fail for 1000s of namespace, what is the better approach for doing migration over 5000 namespaces.

**Comment by dhalperi** [comment]
Hi @chandra-vs,

What have you tried so far, and where are you running into issues? If there are 1000s of namespaces, the solution should work – only apply say 100 namespaces per Dataflow job, and run several (10s) of Dataflow jobs.

Thanks!

**Comment by rajesh99** [comment]
Can the Dataflow team extend the WordCount example to handle multiple namespaces.

**Comment by davorbonaci** [comment]
We don't like extending the WordCount examples. Those are getting started examples, and we'd like to keep them as simple as possible.

Do you have a specific question related to a solution proposed by @dhalperi? If you share a code snippet we might be able to help you faster.

**Comment by rajesh99** [comment]
Dear Davor,
I was not able to put my attention back into the namespaces dataflow
issue.  It would help if a direct example is given. Seems like, more
expertise is required.

thanks
Rajesh


On Tue, Jan 26, 2016 at 4:40 AM, Davor Bonaci <notifications@github.com>
wrote:

> Closed #83
> <https://github.com/GoogleCloudPlatform/DataflowJavaSDK/issues/83>.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/GoogleCloudPlatform/DataflowJavaSDK/issues/83#event-525718951>
> .
>



-- 
Rajesh
*www.VeersoftSolutions.com <http://www.VeersoftSolutions.com>*
*www.AccountingGuru.net <http://www.AccountingGuru.net>*
*Accounting/Inventory/Orders on Google Cloud Platform and Mobile*

**Comment by davorbonaci** [comment]
The code posted by @dhalperi already handles this. You just have to fill in the blanks with your custom processing logic.
